header/p2p/session.go

```diff
@@ -0,0 +1,226 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/header"
+	p2p_pb "github.com/celestiaorg/celestia-node/header/p2p/pb"
+)
+
+// TODO(@vgonkivs): make it configurable
+var (
+	// headersPerPeer is a maximum amount of headers that will be requested per peer.
+	headersPerPeer uint64 = 64
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peers with records that describes their activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	ctx, cancel := context.WithCancel(ctx)
+	return &session{
+		ctx:        ctx,
+		cancel:     cancel,
+		protocolID: protocolID,
+		host:       h,
+		queue:      newPeerQueue(ctx, peerTracker),
+		errCh:      make(chan error),
+	}
+}
+
+// GetRangeByHeight requests headers from different peers.
+func (s *session) getRangeByHeight(ctx context.Context, from, amount uint64) ([]*header.ExtendedHeader, error) {
+	log.Debugw("requesting headers", "from", from, "to", from+amount)
+
+	requests := prepareRequests(from, amount, headersPerPeer)
+	result := make(chan []*header.ExtendedHeader, len(requests))
+	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
+
+	go s.handleOutgoingRequests(ctx, result)
+	for _, req := range requests {
+		s.reqCh <- req
+	}
+
+	headers := make([]*header.ExtendedHeader, 0, amount)
+	for i := 0; i < cap(result); i++ {
+		select {
+		case <-s.ctx.Done():
+			return nil, errors.New("header/p2p: exchange is closed")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-s.errCh:
+			return nil, err
+		case res := <-result:
+			headers = append(headers, res...)
+		}
+	}
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Height < headers[j].Height
+	})
+	return headers, nil
+}
+
+// close stops the session.
+func (s *session) close() {
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+}
+
+// handleOutgoingRequests pops a peer from the queue and sends a prepared request to the peer.
+func (s *session) handleOutgoingRequests(ctx context.Context, result chan []*header.ExtendedHeader) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.ctx.Done():
+			return
+		case req := <-s.reqCh:
+			stats := s.queue.waitPop(ctx)
+			if stats.peerID == "" {
+				return
+			}
+			go s.doRequest(ctx, stats, req, result)
+		}
```

Comment:
I'm sorta worried that `handleOutgoingRequests` hangs longer that it needs to and we have no way to know when it is no longer needed. The only way to exit for loop is by one of the contexts being canceled. Do we expect the number of hanging routines to be low?

What do you think about  passing `requests` slice directly to this function, so it could help to know ahead of time required amount of iterations. Then `s.reqCh` could be removed. moving retry logic inside of `s.doRequest` by spinning new retry goroutine `go s.doRequest` directly from there instead of communicating it back by `s.reqCh`?

Also if waitPop is turned into single channel (see my other suggestion) it could be used directly from select.

Reply:
> handleOutgoingRequests hangs longer that it needs to and we have no way to know when it is no longer needed

We are stopping it when all headers have been received.

> so it could help to know ahead of time required amount of iterations

We do not know the exact amount of iterations with retrying logic.

Reply:
Had a call with @walldiss and explained how it supposed to work.

Reply:
Thank you again for explanation. Could you please add few comment for places in code that were difficult for me, so it is easier others:
-  `waitPop` selects peer with highest score among the available ones for the request 
- `handleOutgoingRequests` goroutine will exit via canceled session context when all request are processed.